core/domain/opportunity_services.py

```diff
@@ -386,3 +405,65 @@ def update_opportunities_with_new_topic_name(topic_id, topic_name):
 
     _save_multi_exploration_opportunity_summary(
         exploration_opportunity_summary_list)
+
+
+def regenerate_opportunities_related_to_topic(
+        topic_id, delete_existing_opportunities=False):
+    """Regenerates opportunity models which belongs to a given topic.
+
+    Args:
+        topic_id: str. The ID of the topic.
+        delete_existing_opportunities: bool. Whether to delete all the existing
+            opportunities related to the given topic.
+
+    Returns:
+        dict. A dict representing the result of the regeneration process.
+    """
+    if delete_existing_opportunities:
+        exp_opportunity_models = (
+            opportunity_models.ExplorationOpportunitySummaryModel.get_by_topic(
+                topic_id))
+        opportunity_models.ExplorationOpportunitySummaryModel.delete_multi(
+            exp_opportunity_models)
+
+    topic = topic_fetchers.get_topic_by_id(topic_id)
+    story_ids = topic.get_canonical_story_ids()
+    stories = story_fetchers.get_stories_by_ids(story_ids)
+    exp_ids = []
+    non_existing_story_ids = []
+
+    for index, story in enumerate(stories):
+        if story is None:
+            non_existing_story_ids.append(story_ids[index])
+        else:
+            exp_ids += story.story_contents.get_all_linked_exp_ids()
+
+    exp_ids_to_exp = exp_fetchers.get_multiple_explorations_by_id(
+        exp_ids, strict=False)
+    non_existing_exp_ids = set(exp_ids) - set(exp_ids_to_exp.keys())
+
+    if len(non_existing_exp_ids) > 0 or len(non_existing_story_ids) > 0:
+        return {
+            'status': 'FAILED',
+            'missing_exp_with_ids': list(non_existing_exp_ids),
```

Comment:
I don't think you need to programmatize this -- it should be fine to convey this info in an error message string. (You don't actually use the raw values directly anywhere, right?)

Reply:
Done!